### PR TITLE
createLocalVue errorHandler Option

### PR DESCRIPTION
**Pull request by AtofStryker**
The purpose of this PR is to allow an [`errorHandler`](https://vuejs.org/v2/api/#errorHandler) option for `createLocalVue`. The goal is to support an interface to allow for support over time as needed for Vue Global Config [options](https://vuejs.org/v2/api/#Global-Config).

``` .js
createLocalVue({
  errorHandler(err, vm, info) {
    // catch errors on local vue instance here
  }
})
```

`createLocalVue` used to take a private argument, which was a vue instance. The public API does not use this `_vue` instance argument, but `server-test-utils` and `test-utils` both do. The first step to supporting this signature was to move `create-local-vue` to the `shared` directory and use the semantic `_createLocalVue` when referencing internally. The "new" `createLocalVue` (part of the public api) just wraps this by passing through `undefined` to get the default behavior as well as the propagated `VueConfig`.

It is worth noting that the VTU behavior for disallowing a global error handler is still valid. The way this feature works is by invoking the registered `errorHandler` in the VTU global error handler before throwing the error. The function traverses the components `$parent` up the tree, finds if a localVue instance exists, and if so, invoke the `errorHandler` option if one is defined. That can be seen [here](https://github.com/vuejs/vue-test-utils/pull/1670/files#diff-2e7522386f80e82b395ed39a29c734fbR10). The user defined `errorHandler` passed into `localVue` can also throw an error if desired, but is not recommended.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue-test-utils/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it): Addresses https://github.com/vuejs/vue-test-utils/issues/1503

**Other information:**

### CODE REVIEW

File: packages/test-utils/src/error.js [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+import { findAllInstances, findAllParentInstances } from './find'
 
-function errorHandler(errorOrString, vm) {
+function errorHandler(errorOrString, vm, info) {
   const error =
     typeof errorOrString === 'object' ? errorOrString : new Error(errorOrString)
 
+  // If a user defined errorHandler was register via createLocalVue
+  // find and call the user defined errorHandler
+  const instancedErrorHandlers = findAllParentInstances(vm)
```
Comment (AtofStryker) [inline packages/test-utils/src/error.js:10]: maybe just move this into line `21` that already has the conditional check? A lot of the type checking is for older versions of Vue

File: packages/test-utils/src/find.js [hunk 2]
```diff
@@ -10,6 +10,26 @@ import {
...
 
+/**
+ * Traverses a vue instance for its parents and returns them in an array format
+ * @param {Component} vm
+ * @returns {Component[]} The component and its corresponding parents, in order from left to right
+ */
+export function findAllParentInstances(childVm: any) {
+  const instances = [childVm]
+
+  function getParent(_vm) {
```
Comment (AtofStryker) [inline packages/test-utils/src/find.js:21]: Any ideas on how to avoid the closure here?

File: packages/shared/create-local-vue.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        instance[key] = original
+      }
+    }
+  })
+
+  // config is not enumerable
+  instance.config = cloneDeep(Vue.config)
+
+  // if a user defined errorHandler is defined by a localVue instance via createLocalVue, register it
+  instance.config.errorHandler = config.errorHandler || Vue.config.errorHandler
```
Comment (AtofStryker) [inline packages/shared/create-local-vue.js:39]: This is the only line that has changed within the logic, IE adding `config.errorHandler`. If we do not feel confortable reviewing this due to scope of change, we can break out the `shared` move in a separate PR to more clearly see the change set

File: packages/shared/create-local-vue.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Used internally by vue-server-test-utils and test-utils to propagate/create vue instances.
+ * This method is wrapped by createLocalVue in test-utils to provide a different public API signature
+ * @param {Component} _Vue
+ * @param {VueConfig} config
+ * @returns {Component}
+ */
+function _createLocalVue(
+  _Vue: Component = Vue,
+  config: VueConfig = {}
```
Comment (AtofStryker) [inline packages/shared/create-local-vue.js:15]: added argument `config: VueConfig = {}`

**Comment by lmiller1990** [comment]
Hi! This is a pretty large PR, please give me the weekend to review. I am excited for this feature.

**Comment by AtofStryker** [comment]
> Hi! This is a pretty large PR, please give me the weekend to review. I am excited for this feature.

@lmiller1990 sure thing! I am excited about it too. Definitely no rush on review as we want to be thorough. 

What we can do, since this is a fairly large PR where the change set is a bit difficult to determine, is break this out into 2 separate PRs:

1st PR (break out `createLocalVue` to `shared` and wrap the functionality around the new `createLocalVue`)

48f1c14
808fe89
fc3be6c
0dca580

2nd (implement the `errorHandler`)
everything else

I think this would be much easier to review as well as add confidence to what we are doing. What do you think?

**Comment by lmiller1990** [comment]
I am happy to review a big PR, no dramas there.

**Comment by AtofStryker** [comment]
@lmiller1990 OK cool! I just want to make it as easy as possible and willing to accommodate :smiley:

**Comment by lmiller1990** [comment]
Sorry about the wait, I have a day off work this week to work on OSS (mainly VTU stuff).

I will review/merge and plan release 1.1 for this weekend.

**Comment by AtofStryker** [comment]
@lmiller1990 no worries! Let me know if there is anything I can do to help.

File: packages/shared/create-local-vue.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  // option merge strategies need to be exposed by reference
+  // so that merge strats registered by plugins can work properly
+  instance.config.optionMergeStrategies = Vue.config.optionMergeStrategies
+
+  // make sure all extends are based on this instance.
+  // this is important so that global components registered by plugins,
+  // e.g. router-link are created using the correct base constructor
+  instance.options._base = instance
+
+  // compat for vue-router < 2.7.1 where it does not allow multiple installs
```
Comment (lmiller1990) [inline packages/shared/create-local-vue.js:50]: nice job, this must have been a huge pain to figure out

Comment (lmiller1990) [reply packages/shared/create-local-vue.js:50]: oh, it was moved here from somewhere else - even better, that means it's been battle tested

File: packages/shared/create-local-vue.js (hunk 3 above)
Comment (lmiller1990) [reply packages/shared/create-local-vue.js:39]: shared is private anyway, this should be fine

File: packages/test-utils/src/find.js (hunk 2 above)
Comment (lmiller1990) [reply packages/test-utils/src/find.js:21]: This is to find the highest `Vue` with the custom `errorHandler` right? I somehow thought it would "just work" and we wouldn't need to reverse-traverse all the way up. I don't have another way to do this. 

I think having the conditional is fine, no problem from my point of view

**Review by lmiller1990 — APPROVED** [review]
Looks fine to me - this was a ton of work! Great job. I wonder if this will be easier to do in VTU next. I was not expecting this to be so complicated (I don't think you were, either).

File: packages/test-utils/src/find.js (hunk 2 above)
Comment (AtofStryker) [reply packages/test-utils/src/find.js:21]: @lmiller1990 correct. I thought it would work straight out of the box as well, but not exactly for reasons I am having trouble remembering. I want to say it has to do with how the vue instances behave on the component, but don't take my word for it 😅

**Comment by AtofStryker** [comment]
> Looks fine to me - this was a ton of work! Great job. I wonder if this will be easier to do in VTU next. I was not expecting this to be so complicated (I don't think you were, either).

@lmiller1990 Thanks! Yeah... this was a lot more involved than I thought it would be 😅 . The first approach I took failed miserably 😆 . This seemed to be the way to get the behavior we need with the least amount of change (even though significant change was needed). I will say this is WAYYYY easier with VTU Next. [This](https://github.com/vuejs/vue-test-utils/issues/1503#issuecomment-679457509) is what it looks like in VTU Next, which works exactly like you would expect it to 🙂 .

**Comment by lmiller1990** [comment]
Oh yeah. Forgot about that thread in VTU next.

Anyway, I released 1.1.0, thanks for your help! Pretty significant release.

**Comment by thibaut-lo** [comment]
Shouldn't `export declare function createLocalVue (): typeof Vue` be updated (https://github.com/vuejs/vue-test-utils/blob/dev/packages/test-utils/types/index.d.ts#L172) ?

**Comment by coolboy0961** [comment]
> Shouldn't `export declare function createLocalVue (): typeof Vue` be updated (https://github.com/vuejs/vue-test-utils/blob/dev/packages/test-utils/types/index.d.ts#L172) ?

Because this, I can not test errorHandler by Typescript.

**Comment by lmiller1990** [comment]
I don't know what you mean, but if you'd like to improve the types, please make a PR. Alternatively, if you don't know how to fix them, you can open an issue  sharing the problem you are having. Thanks!

**Comment by coolboy0961** [comment]
@lmiller1990 
Thank you for the response. I added a issue.
https://github.com/vuejs/vue-test-utils/issues/1872

I think it just need to fix the type file, I wanted to make a PR,
but I do not have the permission to push the fix branch.

**Comment by coolboy0961** [comment]
@lmiller1990 
I made a PR
https://github.com/vuejs/vue-test-utils/pull/1873

It turned out to be Fork first and then create a pull request.
